### ISSUE TITLE
Improve Sink.combine to return materialized values of all Sinks

### ISSUE THREAD

**Issue by markusthoemmes** [akka/akka#23958]
The current implementation of `Sink.combine` materializes to `NotUsed` for quite obvious reasons: There is no way of getting the materialization types of the `rest` parameter.

Since `Sink.combine` is a super handy method to easily create FanOut/FanIn etc. streams, I kinda expected it to give me the materialized values of all sinks in a tuple.

In my specific case, I implemented that bit myself in:

```scala
def combine[T, U, M1, M2](first: Sink[U, M1], second: Sink[U, M2])(
  strategy: Int ⇒ Graph[UniformFanOutShape[T, U], NotUsed]): Sink[T, (M1, M2)] = {
  Sink.fromGraph(GraphDSL.create(first, second)((_, _)) { implicit b => (s1, s2) =>
    import GraphDSL.Implicits._
    val d = b.add(strategy(2))

    d ~> s1
    d ~> s2

    SinkShape(d.in)
  })
}
```

I can imagine the overrides one would need to be relatively low. `combine20` would probably be overkill and arguably, having `combine2` one can just reapply that method to get a huge stacked tuple back in the end.

In any case: Do you think this would be a useful thing to have?

**Comment by johanandren** [comment]
To follow convention in other factories it should be `combineMat(Sink[U, M1], Sink[U, M2])((M1, M2) => M3): Sink[U, M3]`. I think it is a good idea.

**Comment by markusthoemmes** [comment]
@johanandren sweet! I'd gladly take the task to code it up and everything. Which arities do you think make sense to implement? Only 2 for a start?

**Comment by johanandren** [comment]
Hmm, looking at the existing `combine` it's a 2-mandatory-sinks plus varargs, you think it would be possible to pull off matching that? 

If we limit the type of the matval it could perhaps be something like `combineMat(first: Sink[U, Mat1], second: Sink[U, Mat2], rest: Sink[U, MatN]*)((Mat1, Mat2, immutable.Seq[MatN]) => MatOut)`?

**Comment by markusthoemmes** [comment]
Investigated a bit:

1. It is possible to build any number of combinations using just `combineMat(2)` by just applying it multiple times to the combined sinks again. Return type will not be exactly pretty, but it'd work. I don't know of the performance implications of having multiple `Broadcast(2)` vs having a single `Broadcast(5)` though.

2. At least from my current implementation your proposal is not possible to implement by using `GraphDSL.create` to import the shapes and combine the materialized values. There simply isn't a variadic version of it as of today. Doesn't mean it's impossible to implement though. I kinda doubt its usefulness though. It feels a bit odd to give 2 specific types and then the rest needs to be all of the same type?

**Comment by johanandren** [comment]
Starting to think that this was the reason there never was a `Sink.combineMat` to start with :)

If you want to broadcast to 6 downstream, I don't think you want to do that by creating 5 broadcast stages.

So maybe using GraphDSL is the answer here rather than creating a weird flow-API for it?

**Comment by markusthoemmes** [comment]
`GraphDSL` certainly is, in general, the answer, definitely.

I expected to be able to do a simple FanOut much easier than that though. At least fanning out to 2 or 3 downstream sinks was actually a use-case I imagined to happen quite often (there are some old issues from @ktoso proposing `broadcast` flow APIs as well).

There is, for completeness, a third version of achieving what I want:

```scala
val sequence = Promise[Seq[String]]
val toSeq = Sink.seq[String].mapMaterializedValue(sequence.completeWith)
val toConsole = Sink.foreach[String](line => println(line))

val combined = Sink.combine(toSeq, toConsole)(Broadcast(_))
...
```

That'd give me the materialized values I want to have, but it introduces promises to my code so.... meh 😁.

Personally, I liked the possibility of having the `Sink`be statically typed with all the values... well... `combine`d together.

**Comment by johanandren** [comment]
Ok, what about doing the original combineMat idea with 2-arity for now, and maybe mention in the scala docs that for higher arities, use GraphDSL?

**Comment by markusthoemmes** [comment]
Sounds good to me @johanandren :)

**Comment by tonicsoft** [comment]
Above, it is suggested this problem can be solved with GraphDSL:

> If you want to broadcast to 6 downstream, I don't think you want to do that by creating 5 broadcast stages.
>
> So maybe using GraphDSL is the answer here rather than creating a weird flow-API for it?

However, as far as I know, there is no way to combine more than 2 materialised values using the GraphDSL. Therefore, even with the graph dsl, you would still have to create 5 broadcast stages to combine all 6 materialised values. Is this correct?

**Comment by johanandren** [comment]
There are `GraphDSL.create` overloads that take up to 22 input graphs and allows to combine their materialized values.

**Comment by tonicsoft** [comment]
Ah that's true, but I was thinking one had a list of sinks of arbitrary or configurable size (say your app is partitioned) rather than 22 hard coded sinks. I.e. I don't want to change the combine code every time I want to change the number of sinks.

**Comment by johanandren** [comment]
There's a PR in progress working on a version that accepts a list of graphs: #24102

**Comment by tonicsoft** [comment]
Ah that's awesome! I've been missing that feature, and looks like it would also take care of this issue. Keen to hear what @markusthoemmes thinks as #24102 looks like it will handle this in a one(ish) liner.

**Comment by markusthoemmes** [comment]
@tonicsoft With #24102 you'd only be able to get a list of materialized values of exactly one type, does it? This proposal is more about returning the distinct value (+ type) of each of the Sinks.

**Comment by tonicsoft** [comment]
You are correct. I was thinking about the `rest` parameter referred to in the first comment (which of course could only have one type), but I see now you agreed to tackle only 2 for now.

**Comment by acxlam** [comment]
@markusthoemmes happy to see this ticket. Could `CompletionStage<Done>`  for the whole graph be more useful than `NotUsed`?

Have just started Akka Streams in some, hopefully I am asking a valid question.

IMHO, It is eaiser for Java programers with RxJava/Reactor experience to start by using the combined sink i/o GraphDSL since the simpler API in Java resembles `ConnectableFlux` much. If the completion of the graph can be observed more easily. That will be nicer.

**Comment by acxlam** [comment]
Found out I can prematerialize before combine. That's cool, hope that's in the document as a means for adding callback to whole graphs.

**Comment by hepin1989** [comment]
I encounter this too... hope I can reduce all the materialized value into a builder and retrive from the built result.

**Comment by hepin1989** [comment]
```scala
 def combine[T, U, M](sources: immutable.Seq[Source[T,_<:M]])(
    strategy: Int => Graph[UniformFanInShape[T, U], NotUsed]): Source[U, Seq[M]] =
    Source.fromGraph(GraphDSL.create(sources) { implicit b => shapes =>
      import GraphDSL.Implicits._
      val c = b.add(strategy(sources.size))
      for((s, i) <- shapes.zipWithIndex) {
        s ~> c.in(i)
      }
      SourceShape(c.out)
    })
```
@johanandren do you think this can be added?

**Comment by He-Pin** [comment]
This issue may require some kind of code generation I think.

**Comment by johanandren** [comment]
Partially fixed by #31345
